File: measure/src/main/scala/org/apache/griffin/measure/data/source/DataSource.scala
/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/
package org.apache.griffin.measure.data.source

import org.apache.griffin.measure.cache.tmst._
import org.apache.griffin.measure.data.connector._
import org.apache.griffin.measure.data.connector.batch._
import org.apache.griffin.measure.data.connector.streaming._
import org.apache.griffin.measure.data.source.cache.OldDataSourceCache
import org.apache.griffin.measure.log.Loggable
import org.apache.griffin.measure.process.temp.{DataFrameCaches, TableRegisters, TimeRange}
import org.apache.griffin.measure.rule.plan.TimeInfo
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.{DataFrame, Row, SQLContext}

case class DataSource(sqlContext: SQLContext,
                      name: String,
                      baseline: Boolean,
                      dataConnectors: Seq[DataConnector],
                      dataSourceCacheOpt: Option[OldDataSourceCache]
                     ) extends Loggable with Serializable {

  val batchDataConnectors = DataConnectorFactory.filterBatchDataConnectors(dataConnectors)
  val streamingDataConnectors = DataConnectorFactory.filterStreamingDataConnectors(dataConnectors)
  streamingDataConnectors.foreach(_.dataSourceCacheOpt = dataSourceCacheOpt)

  val tmstCache: TmstCache = TmstCache()

  def init(): Unit = {
    dataSourceCacheOpt.foreach(_.init)
    dataConnectors.foreach(_.init)

    dataSourceCacheOpt.map(_.tmstCache = tmstCache)
    dataConnectors.map(_.tmstCache = tmstCache)
  }

  def loadData(timeInfo: TimeInfo): TimeRange = {
    val calcTime = timeInfo.calcTime
    println(s"load data [${name}]")
    val (dfOpt, tmsts) = data(calcTime)
    dfOpt match {
      case Some(df) => {
//        DataFrameCaches.cacheDataFrame(timeInfo.key, name, df)
        TableRegisters.registerRunTempTable(df, timeInfo.key, name)
      }
      case None => {
        warn(s"load data source [${name}] fails")
      }
    }
    tmsts
  }

  private def data(ms: Long): (Option[DataFrame], TimeRange) = {
    val batches = batchDataConnectors.flatMap { dc =>
      val (dfOpt, timeRange) = dc.data(ms)
      dfOpt match {
        case Some(df) => Some((dfOpt, timeRange))
        case _ => None
      }
    }
    val caches = dataSourceCacheOpt match {
      case Some(dsc) => dsc.readData() :: Nil
      case _ => Nil
    }
    val pairs = batches ++ caches

    if (pairs.size > 0) {
      pairs.reduce { (a, b) =>
        (unionDfOpts(a._1, b._1), a._2.merge(b._2))
      }
    } else {
      (None, TimeRange.emptyTimeRange)
    }
  }

  private def unionDfOpts(dfOpt1: Option[DataFrame], dfOpt2: Option[DataFrame]
                         ): Option[DataFrame] = {
    (dfOpt1, dfOpt2) match {
      case (Some(df1), Some(df2)) => Some(unionDataFrames(df1, df2))
      case (Some(df1), _) => dfOpt1
      case (_, Some(df2)) => dfOpt2
      case _ => None
    }
  }

  private def unionDataFrames(df1: DataFrame, df2: DataFrame): DataFrame = {
    try {
      val cols = df1.columns
      val rdd2 = df2.map{ row =>
        val values = cols.map { col =>
          row.getAs[Any](col)
        }
        Row(values: _*)
      }
      val ndf2 = sqlContext.createDataFrame(rdd2, df1.schema)
      df1 unionAll ndf2
    } catch {
      case e: Throwable => df1
    }
  }

  def updateData(df: DataFrame, ms: Long): Unit = {
    dataSourceCacheOpt.foreach(_.updateData(df, ms))
  }

  def updateDataMap(dfMap: Map[Long, DataFrame]): Unit = {
    dataSourceCacheOpt.foreach(_.updateDataMap(dfMap))
  }

  def cleanOldData(): Unit = {
    dataSourceCacheOpt.foreach(_.cleanOldData)
  }

}

File: service/src/main/java/org/apache/griffin/core/measure/entity/DataSource.java
/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/

package org.apache.griffin.core.measure.entity;


import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.JoinColumn;
import javax.persistence.OneToMany;
import javax.persistence.PostLoad;
import javax.persistence.PrePersist;
import javax.persistence.PreUpdate;
import javax.persistence.Transient;

import org.apache.griffin.core.util.JsonUtil;
import org.springframework.util.CollectionUtils;
import org.springframework.util.StringUtils;

@Entity
public class DataSource extends AbstractAuditableEntity {
    private static final long serialVersionUID = -4748881017079815794L;

    private String name;

    @OneToMany(fetch = FetchType.EAGER, cascade = {CascadeType.PERSIST,
            CascadeType.REMOVE, CascadeType.MERGE})
    @JoinColumn(name = "data_source_id")
    private List<DataConnector> connectors = new ArrayList<>();

    private boolean baseline = false;

    @JsonIgnore
    @Column(length = 1024)
    private String checkpoint;

    @Transient
    @JsonInclude(JsonInclude.Include.NON_NULL)
    private Map<String, Object> checkpointMap;


    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public List<DataConnector> getConnectors() {
        return connectors;
    }

    public void setConnectors(List<DataConnector> connectors) {
        if (CollectionUtils.isEmpty(connectors)) {
            throw new NullPointerException("Data connector can not be empty.");
        }
        this.connectors = connectors;
    }

    public boolean isBaseline() {
        return baseline;
    }

    public void setBaseline(boolean baseline) {
        this.baseline = baseline;
    }

    private String getCheckpoint() {
        return checkpoint;
    }

    private void setCheckpoint(String checkpoint) {
        this.checkpoint = checkpoint;

    }

    @JsonProperty("checkpoint")
    public Map<String, Object> getCheckpointMap() {
        return checkpointMap;
    }

    public void setCheckpointMap(Map<String, Object> checkpointMap) {
        this.checkpointMap = checkpointMap;
    }

    @PrePersist
    @PreUpdate
    public void save() throws JsonProcessingException {
        if (checkpointMap != null) {
            this.checkpoint = JsonUtil.toJson(checkpointMap);
        }
    }

    @PostLoad
    public void load() throws IOException {
        if (!StringUtils.isEmpty(checkpoint)) {
            this.checkpointMap = JsonUtil.toEntity(
                checkpoint, new TypeReference<Map<String, Object>>() {
                });
        }
    }

    public DataSource() {
    }

    public DataSource(String name, List<DataConnector> connectors) {
        this.name = name;
        this.connectors = connectors;
    }

    public DataSource(String name, boolean baseline,
                      Map<String, Object> checkpointMap,
                      List<DataConnector> connectors) {
        this.name = name;
        this.baseline = baseline;
        this.checkpointMap = checkpointMap;
        this.connectors = connectors;

    }
}
